[lldb] Make GetDynamicTypeAndAddress() use host address if available

augusto2112 opened the pull request:
[lldb] Make GetDynamicTypeAndAddress() use host address if available

    If GetDynamicTypeAndAddress_Value cannot get back to the inferior address,
    but does have a host address and buffer that fits the dynamic type,
    return that.


    [lldb] Make ValueObjectDynamicValue::UpdateValue() point to  a host b… (#125143)

    …uffer

    ValueObjectDynamicValue::UpdateValue() assumes that the dynamic type
    found by GetDynamicTypeAndAddress() would return an address in the
    inferior. This commit makes it so it can deal with being passed a host
    address instead.

    This is needed downstream by the Swift fork.

    rdar://143357274
    (cherry picked from commit 0cbc4983adcdbbd85ccb38b2dfbfe5985367b1b2)

augusto2112 committed:
[lldb] Make ValueObjectDynamicValue::UpdateValue() point to  a host b… (#125143)

…uffer

ValueObjectDynamicValue::UpdateValue() assumes that the dynamic type
found by GetDynamicTypeAndAddress() would return an address in the
inferior. This commit makes it so it can deal with being passed a host
address instead.

This is needed downstream by the Swift fork.

rdar://143357274
(cherry picked from commit 0cbc4983adcdbbd85ccb38b2dfbfe5985367b1b2)

augusto2112 committed:
[lldb] Make GetDynamicTypeAndAddress() use host address if available

If GetDynamicTypeAndAddress_Value cannot get back to the inferior address,
but does have a host address and buffer that fits the dynamic type,
return that.

augusto2112 commented:
@adrian-prantl one commit is just a cherry-pick. The second commit is the one that needs review.

adrian-prantl commented on lldb/source/ValueObject/ValueObjectDynamicValue.cpp:
```diff
@@ -277,11 +278,29 @@ bool ValueObjectDynamicValue::UpdateValue() {
-    lldb::addr_t load_address = m_address.GetLoadAddress(target_sp.get());
-    m_value.GetScalar() = load_address;
+    // If we found a host address, and the dynamic type fits in the local buffer
+    // that was found, point to thar buffer. Later on this function will copy
```
typo: `that`

augusto2112 commented:
@swift-ci test

adrian-prantl commented on lldb/source/ValueObject/ValueObjectDynamicValue.cpp:
```diff
@@ -277,11 +278,29 @@ bool ValueObjectDynamicValue::UpdateValue() {
+      if (local_buffer.size() <
+          m_dynamic_type_info.GetCompilerType().GetByteSize(exe_scope)) {
+        SetValueIsValid(false);
+        return false;
```
It would be nice at some point for this to return an error instead of a bool so we could communicate the failure cause.

adrian-prantl commented on lldb/source/Plugins/LanguageRuntime/Swift/SwiftLanguageRuntimeDynamicTypeResolution.cpp:
```diff
@@ -2430,17 +2432,41 @@ bool SwiftLanguageRuntime::GetAbstractTypeName(StreamString &name,
+
+    // If we can't find the size of the local buffer we can't safely know if the
+    // dynamic type fits in it.
+    if (local_buffer_size == LLDB_INVALID_ADDRESS)
```
This would be cleaner if in_value.GetLocalBufferSize() returned an optional

adrian-prantl replied:
Or even better start+size should be a StringRef (or MemoryBuffer)

adrian-prantl commented on lldb/source/Plugins/LanguageRuntime/Swift/SwiftLanguageRuntimeDynamicTypeResolution.cpp:
```diff
@@ -2510,7 +2536,8 @@ void SwiftLanguageRuntime::DumpTyperef(CompilerType type,
     ValueObject &in_value, CompilerType dynamic_type,
-    Value::ValueType static_value_type, bool is_indirect_enum_case) {
+    Value::ValueType static_value_type, bool is_indirect_enum_case,
+    llvm::ArrayRef<uint8_t> &local_buffer) {
```
ArrayRefs are meant to be only passed by value. Is this an out parameter?

augusto2112 replied:
I guess this particular function doesn't need the reference.

Commit:
[lldb] Fix build error in ValueObject test

Fixes 0cbc4983adcdbbd85ccb38b2dfbfe5985367b1b2.

llvm-project/lldb/unittests/ValueObject/DynamicValueObjectLocalBuffer.cpp(221): error C2065: 'u_int8_t': undeclared identifier

(cherry picked from commit 9d83790d170cc0a06179514e2ab739f7ecafb78c)

Commit:
[lldb] Fix compiler error in ValueObject tests

Fixes 0cbc4983adcdbbd85ccb38b2dfbfe5985367b1b2.

error: non-constant-expression cannot be narrowed from type 'uint64_t' (aka 'unsigned long long') to 'size_t' (aka 'unsigned int') in initializer list [-Wc++11-narrowing]
   71 |                     in_value.GetLocalBufferSize()};
      |                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Given this is test code, I think a static cast is fine here.

Tiny risk that it's actually a symptom of a bug that would
happen if you did 32-bit to 64-bit debugging. I expect you'd
find a lot more bugs than that though.

(cherry picked from commit 52db30ec4154b0ef21db546ed9b5a9bfe47859cd)

augusto2112 committed:
[lldb][NFC] Replace GetLocalBufferSize() with GetLocalBuffer()

GetLocalBuffer() makes more sense as an API.

augusto2112 committed:
[lldb][NFC] Replace usage of GetLocalBufferSize with GetLocalBuffer

augusto2112 commented:
@swift-ci test

augusto2112 commented:
Windows failures are on the compiler side which shouldn't be related to these changes:
```
  Swift(windows-x86_64) :: Interop/Cxx/stdlib/use-std-span.swift
  Swift(windows-x86_64) :: Interop/Cxx/stdlib/use-std-string-view.swift
```
